pkg/event/emitter.go

```diff
@@ -44,38 +44,57 @@ type Interface interface {
 type Emitter struct {
 	*loader
 
-	log            *zap.SugaredLogger
-	listeners      common.Listeners
-	listenerExists map[string]map[string]struct{}
-	mutex          sync.RWMutex
-	bus            bus.Bus
-	instanceId     string
-	leaseBackend   leasebackend.Repository
-	subjectRoot    string
-	clusterName    string
+	log                 *zap.SugaredLogger
+	registeredListeners common.Listeners
+	listeners           common.Listeners
+	mutex               sync.RWMutex
+	bus                 bus.Bus
+	instanceId          string
+	leaseBackend        leasebackend.Repository
+	subjectRoot         string
+	clusterName         string
 }
 
-// appendUniqueListeners appends only listeners unique by kind and name.
-func (e *Emitter) appendUniqueListeners(listeners ...common.Listener) {
-	e.mutex.Lock()
-	defer e.mutex.Unlock()
+// uniqueListeners keeps a unique set of listeners by kind and name.
+// The last listeners for each kind and name combination takes precedence.
+func uniqueListeners(listeners []common.Listener) []common.Listener {
+	exists := make(map[string]map[string]common.Listener)
+
+	// Determine latest version for each listener by kind and name
 	for i := range listeners {
 		kind := listeners[i].Kind()
 		name := listeners[i].Name()
-		if e.listenerExists[kind] == nil {
-			e.listenerExists[kind] = make(map[string]struct{})
+		if exists[kind] == nil {
```

Comment:
You want map access checking here:

```suggestion
		if _, ok := exists[kind]; !ok {
```
normally I would use the variable name `exists` here rather than `ok`, but that won't work nicely because the map of maps is called that.